### DISCUSSION TITLE
Exclude code when running in a test

### DISCUSSION THREAD

**Discussion by Symbai** [xunit/xunit#2530] [Question] [answered]
In my xunit project I'm accessing a static class of my application. In this static class there are many properties which are irrelevant for the test but they're getting initialized and xunit fails. I don't want to rewrite everything only for getting xunit to work.

Are there any kind of preprocessor directives I can use to simply exclude these properties when running in xunit only? Something like:

```C#
#if !xunit
    internal static readonly BitmapImage ErrorImage = new(new Uri("pack://application:,,,/Images/Error.jpg"));
#endif
```

**Comment by bradwilson** [comment, marked as answer]
We don't emit any compiler directives during builds (and it's outside the scope of what we do).

Assuming you're building the code twice (once for production and once for testing), you could do this yourself by defining this only during the test builds. If you're not building it twice, then I'd suggest you start doing so, because the only way to get this kind of compilation change is by building the code twice.

Alternatively, you could leverage [`Lazy<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1) to stop running all your static startup code until the first time the value in question is needed.

**Reply by Symbai** [reply]
> Assuming you're building the code twice

Not sure what you mean by that. I have a project and added another project with all the xunit tests and added a reference to the original project. I am not building anything. I run the tests in VS and that's it.

There has to be a way to exclude code from being part of the tests, without modifying how the application works (so Lazy is not an option). Preprocessor directives are what I would have thought of, because they are exactly for this purpose.

**Reply by bradwilson** [reply]
Preprocessor directives would work fine, but that implies you're building the code-under-test twice: once for production usage, and once for test usage. There's no other way for a preprocessor directive to work (since it only works at compile time, not at runtime).

**Reply by Symbai** [reply]
Yes but when I run the tests through VS the preprocessor directives for the test could kick in. And when I just build / debug / publish my app they don't. I mean VS knows that I am running the tests and when I dont.

**Reply by bradwilson** [reply]
If that's what you want, then you should be asking the Visual Studio team for this feature, not us.
